indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java

```diff
@@ -514,7 +522,14 @@ public TaskActionSegmentPublisher(Task task, TaskToolbox taskToolbox)
     @Override
     public void publishSegment(DataSegment segment) throws IOException
     {
-      taskToolbox.publishSegments(ImmutableList.of(segment));
+      if (taskToolbox.getTaskActionClient().submit(new SetLockCriticalStateAction(segment.getInterval(), TaskLockCriticalState.UPGRADE))) {
+        taskToolbox.publishSegments(ImmutableList.of(segment));
+      } else {
+        throw new ISE(
+            "WTF?! Lock upgrade failed for interval [%s] for RealtimeIndex Task !! This is really bad !! Is there a higher priority task running ?",
+            segment.getInterval()
+        );
```

Comment:
If this exception is thrown that means the lock was not upgraded and will be released when the task finishes. Does this answer your question ?